File: HackerRank/DataStructures/LinkedLists/ReverseLinkedList.java
static SinglyLinkedListNode reverse(SinglyLinkedListNode head) {
    if (head == null){
        return head;
    }
    SinglyLinkedListNode prev = head;
    SinglyLinkedListNode cur = head.next;
    SinglyLinkedListNode nxt = cur.next;
    prev.next = null;
    while(nxt != null){
        cur.next = prev;
        prev = cur;
        cur = nxt;
        nxt = nxt.next;
    }
    head = cur;
    head.next = prev;
    return head;
}

File: Interview Bit/Linked Lists/reverse_linked_list.cpp
/*
* https://www.interviewbit.com/problems/reverse-linked-list/
* Reverse a linked list. Do it in-place and in one-pass.
* For example:
* Given 1->2->3->4->5->NULL,
* return 5->4->3->2->1->NULL.
*
* mycodeschool:
* https://youtu.be/sYcOK51hl-A
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* Solution::reverseList(ListNode* A) {
    struct ListNode* current;
    struct ListNode* prev = NULL;
    struct ListNode* next;
    current = A;
    while(current != NULL){
        next = current -> next;
        current -> next = prev;
        prev = current;
        current = next;
    }
    A = prev;
    return A;
}
